src/list/mod.rs

```diff
@@ -0,0 +1,513 @@
+use crate::allocator::{Allocator, DefaultAllocator};
+use crate::list::node::{ListNode, ListNodeBase};
+use moveit::{new, New};
+use std::marker::PhantomData;
+use std::mem::size_of;
+use std::{fmt, ptr};
+
+mod node;
+
+/// List with the default allocator.
+pub type DefaultList<V> = List<V, DefaultAllocator>;
+
+/// A doubly linked list.
+/// The API is modelled after `std::collections::LinkedList`.
+#[repr(C)]
+pub struct List<T, A: Allocator> {
+    /// Sentinel node, contains the front and back node pointers (prev = back, next = front)
+    node: ListNodeBase<T>,
+    size: u32,
+    allocator: A,
+}
+
+impl<T, A: Allocator> List<T, A> {
+    /// Create a new, empty list.
+    ///
+    /// # Arguments
+    /// `allocator`: The allocator to use
+    ///
+    /// # Safety
+    /// The resulting list must not be moved.
+    pub unsafe fn new_in(allocator: A) -> impl New<Output = Self> {
+        new::of(Self {
+            node: ListNodeBase::default(),
+            size: 0,
+            allocator,
+        })
+        .with(|this| {
+            let this = this.get_unchecked_mut();
+            this.init_sentinel_node();
+        })
+    }
+
+    // Init list sentinel node
+    fn init_sentinel_node(&mut self) {
+        self.node.prev = &mut self.node;
+        self.node.next = &mut self.node;
+    }
+
+    /// Get the list's size
+    pub fn size(&self) -> usize {
+        self.size as usize
+    }
+
+    /// If the list is empty or not
+    pub fn empty(&self) -> bool {
+        self.size() == 0
+    }
+
+    unsafe fn create_node(&mut self, value: T) -> *mut ListNode<T> {
+        let node = unsafe {
+            self.allocator
+                .allocate::<ListNode<T>>(size_of::<ListNode<T>>())
+                .as_mut()
+        }
+        .unwrap();
+        ptr::write(node.value_mut(), value);
+        node
+    }
+
+    /// Drops the given node
+    unsafe fn drop_node(&mut self, node: *mut ListNodeBase<T>) {
+        // Drop the value
+        ptr::drop_in_place(&mut (*(node as *mut ListNode<T>)).value);
+        // Deallocate the memory for the node
+        self.allocator.deallocate(node, size_of::<ListNode<T>>());
+    }
+
```

Comment:
Seems like a bit of an over-optimization to me, why not use `remove_node`?

Reply:
I'd agree that not using `remove_node` for the `Drop` implementation of the list just for using `drop_in_place` instead of `read` would be over-optimisation, however, I think not unnecessarily rewriting all these pointers (as would happen w/ calling `remove_node` atm) is a good optimisation to make.

Reply:
I have just inlined that code into `Drop` now.